### ISSUE TITLE
remote: efficiently collect directories

### ISSUE THREAD

**Issue by mroutis** [iterative/dvc#2648]
**Version**: 0.62.1

**Description**: The current implementation for `_collect_dir` is an **N+1** operation, where it `walk`s the directory to list all the files and then for each one compute/request its checksum.

https://github.com/iterative/dvc/blob/4171aac0294fd316d51558d2593d10ff006221c2/dvc/remote/base.py#L195-L231

The `state` saves us from recomputing/requesting all the checksums again (the `N` operation).
There are remotes like `S3` that have an operation to list the objects with their ETags (`list_objects`).

For me, it make sense to take advantage of this operation, and replace the N+1 (`walk` + `get_file_checksum` * _not-in-state files_) with the one that returns the list of files with some metadata already (`list_objects` in case of S3)

**Comment by shcheklein** [comment]
@mroutis again, could you please provide an example when it can benefit us? Am I correct that it will be a custom implementation for S3 to calculate a directory hash, for example when we need to use it as an external dependency and/or output? 

If that is the case, does it mean, for example that it can be 1 vs N api calls via boto3? do we have some numbers to compare? It's hard to make any decisions w/p any context.

**Comment by mroutis** [comment]
Yes, @shcheklein , this would be a custom implementation for S3.
I'll provide a more concrete example with some numbers when the support for directories on S3 is fully implemented (#2619)

**Comment by mroutis** [comment]
I would leave this here:
```python
    def _collect_dir(self, path_info):
        # See: https://github.com/iterative/dvc/issues/2648
        root = path_info.path

        return [
            {
                self.PARAM_CHECKSUM: entry["ETag"].strip('"'),
                self.PARAM_RELPATH: os.path.relpath(entry["Key"], start=root),
            }
            for entry in self._list_objects(path_info)
        ]
```

It is the implementation I was thinking for S3

**Comment by efiop** [comment]
Sounds good @mroutis ! It would be interesting to see how much performance we are getting from overriding `_collect_dir` like that.

**Comment by efiop** [comment]
Stale
